obs-ffmpeg: Add HLS threading and dropped frames logic

mv2731 opened the pull request:
Add HLS threading and dropped frames logic by:

- Implementing the consumer-producer paradigm with one thread writing 
to a dynamic array, and one thread reading from it to write to the pipe
- Separating out the HLS functions from the ordinary ffmpeg_mux functions
- Creating 3 new files, obs-ffmpeg-hls-mux.c and two .h files to make functions
accessible between obs-ffmpeg-hls-mux and obs-ffmpeg-mux
-  Implementing frame dropping from the dynamic array if a certain threshold
is exceeded by a packet 
- Adapting the stream key hiding code to the HLS specific file

mv2731 committed:
obs-ffmpeg: Implement producer-consumer paradigm

mv2731 committed:
obs-ffmpeg: Add multithreading logic

mv2731 committed:
obs-ffmpeg: Made methods non-HLS specific

The consumer-producer paradigm should be useful for
any protocol relying on ffmpeg; it need not be HLS
specific. Change functinos that were HLS specific to
general and write a function to write encoded packets
to the dynamic array.

mv2731 committed:
obs-ffmpeg: Add log statements to find segfault

mv2731 committed:
obs-ffmpeg: Modify two threads into one

mv2731 committed:
obs-ffmpeg: Resolve segfault caused by thread args

mv2731 committed:
obs-ffmpeg: Resolve write_thread blocking issue

mv2731 committed:
obs-ffmpeg: Add print statements to debug

mv2731 committed:
obs-ffmpeg: Include packet_ref and release

mv2731 committed:
obs-ffmpeg: Use mux_thread for hls pipe write

mv2731 committed:
obs-ffmpeg: Fix freeing of nonallocated ptr

mv2731 committed:
obs-ffmpeg: Clean up ffmpeg-mux

mv2731 committed:
obs-ffmpeg: Remove unnecessary logic

mv2731 committed:
obs-ffmpeg: Fix encoder_packet_ref

mv2731 committed:
obs-ffmpeg: Add helpful logging statements

mv2731 committed:
obs-ffmpeg: Add log statements

mv2731 committed:
obs-ffmpeg: Add threading_buffer field

Add boolean threading_buffer field to ffmpeg_muxer
so that functions which originally wrote directly to
pipe can write to the dynamic array instead.

mv2731 committed:
obs-ffmpeg: Continue separating hls logic

mv2731 committed:
obs-ffmpeg: Remove print statements

mv2731 committed:
libobs: Revert changes to obs-encoder.c

mv2731 committed:
obs-ffmpeg: Clean up encoder and obs-ffmpeg-mux.c

mv2731 committed:
obs-ffmpeg: Start fixing deactivate

mv2731 committed:
obs-ffmpeg: Write hls-specific deactivate

mv2731 committed:
obs-ffmpeg: Code compiles

mv2731 committed:
obs-ffmpeg: Fix pipe write HLS

mv2731 committed:
obs-ffmpeg: Add comment about destroy

mv2731 committed:
obs-ffmpeg: Added .h files to enable drop frames

mv2731 committed:
obs-ffmpeg: Resolve deadlock

mv2731 committed:
obs-ffmpeg: Add function for writing video pkt

mv2731 committed:
obs-ffmpeg: Change where mutex locks

mv2731 committed:
obs-ffmpeg: Create a new file for obs-ffmpeg-hls-mux.c

mv2731 committed:
obs-ffmpeg: Remove congestion-ctrl from CMakeLists

mv2731 committed:
obs-ffmpeg: Add in obs-ffmpeg-hls-mux

mv2731 committed:
obs-ffmpeg: Delete congestion-control

mv2731 committed:
obs-ffmpeg: Edit .h files and fix includes

mv2731 committed:
obs-ffmpeg: Ensure fxn signatures do not conflict

mv2731 committed:
obs-ffmpeg: Working on mpegts error

mv2731 committed:
obs-ffmpeg: Add bsf:v flag

mv2731 committed:
obs-ffmpeg: Remove set bsf to ffmpeg

mv2731 committed:
obs-ffmpeg: Resolve MPEGTS error

mv2731 committed:
obs-ffmpeg: Delete redundant code

mv2731 committed:
obs-ffmpeg: Fix formatting

mv2731 committed:
obs-ffmpeg: Edit ffmpeg_muxer struct

mv2731 committed:
obs-ffmpeg: Remove unnecessary fxns

mv2731 committed:
obs-ffmpeg: Edit deactivate

mv2731 committed:
obs-ffmpeg: Add in hls_deactivate

mv2731 committed:
obs-ffmpeg: Cleaning up PR

mv2731 committed:
obs-ffmpeg: Remove comments

mv2731 committed:
obs-ffmpeg: Clean up hls-mux fxns

mv2731 committed:
obs-ffmpeg: Revamp stream key logic

mv2731 committed:
obs-ffmpeg: Use proper deactivate

mv2731 committed:
obs-ffmpeg: Remove print statements

mv2731 committed:
obs-ffmpeg: Clean up obs-ffmpeg-hls-mux.c

mv2731 committed:
obs-ffmpeg: Remove unnecessary hls-deactivate

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+#include <obs-module.h>
```
Order includes alphabetically.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+#include "obs-ffmpeg-hls-mux.h"
+
+#define do_log(level, format, ...)                  \
+	blog(level, "[ffmpeg muxer: '%s'] " format, \
```
ffmpeg hls muxer

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+	size_t i;
+
+	pthread_mutex_lock(&stream->write_mutex);
+	for (i = 0; i < stream->mux_packets.num; i++) {
```
for (size_t i = 0; i < stream->mux_packets.num; i++)

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+		stream->mux_thread_joinable = false;
+	}
+
+	struct encoder_packet *packet;
```
Move this inside the for loop.

ushadow reviewed (changes requested):
Please see comments.

mv2731 committed:
obs-ffmpeg: Add hls_deactivate in signal_failure

mv2731 committed:
obs-ffmpeg: Ran clang-format

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+
+int hls_deactivate(struct ffmpeg_muxer *stream, int code)
+{
+	int ret = -1;
```
Is it possible to call deactivate from obs-ffmepg-mux here, and then add the additional steps needed for HLS?

mv2731 replied:
Done.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+
+	if (stream) {
+		if (stream->mux_thread_joinable)
+			pthread_join(stream->mux_thread, NULL);
```
Seems unnecessary since hls_deactivate also calls pthread_join.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+	if (os_sem_init(&stream->write_sem, 0) != 0)
+		goto fail;
+
+	if (obs_output_get_flags(output) & OBS_OUTPUT_SERVICE)
+		stream->is_network = true;
```
This is not needed for HLS.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+	return stream;
+
+fail:
+	pthread_mutex_destroy(&stream->write_mutex);
+	os_sem_destroy(stream->write_sem);
+	os_event_destroy(stream->stop_event);
+	bfree(stream);
```
Call ffmpeg_hls_mux_destroy?

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+	remaining = stream->mux_packets.num;
+	if (remaining) {
+		packet = &stream->mux_packets.array[0];
+		write_packet(stream, packet);
```
Need to return the result of write_packet. Can be just a bool instead of int.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+		if (os_event_try(stream->stop_event) == 0)
+			break;
+
+		int ret = process_packet(stream);
```
process_packet doesn't really return these different status. Can be just a bool.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+	dstr_copy(&stream->printable_path, path_str);
+	info("Writing to path '%s'...", stream->printable_path.array);
+
+	stream->mux_thread_joinable = pthread_create(&stream->mux_thread, NULL,
```
Move this above os_atomic_set_bool(&stream->active, true); Return false if pthread_create failed.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+
+	pthread_mutex_lock(&stream->write_mutex);
+
+	if (stream->active) {
```
active(stream)

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+static void *write_thread(void *data)
+{
+	struct ffmpeg_muxer *stream = data;
+	stream->active = true;
```
Not needed.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,374 @@
+	int64_t buffer_duration_usec;
+	int priority = pframes ? OBS_NAL_PRIORITY_HIGHEST
+			       : OBS_NAL_PRIORITY_HIGH;
+	int64_t drop_threshold = 2 * stream->keyframes;
```
drop threshold should be set based on key frame interval not keyframes. keyframes is number of keyframes. You need to add another field to the ffmpeg_muxer struct to track the key frame interval.

mv2731 committed:
obs-ffmpeg: Respond to first round of review

ushadow commented:
Automatic checks show that plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c is not formatted properly.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,339 @@
+#include "obs-ffmpeg-mux.h"
```
Automatic checks show that plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c is not formatted properly.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -25,6 +25,8 @@
 #include <util/threading.h>
 #include "ffmpeg-mux/ffmpeg-mux.h"
+#include "obs-ffmpeg-mux.h"
+#include "obs-ffmpeg-hls-mux.h"
```
Order includes in alphabetic order.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,339 @@
+
+bool hls_deactivate(struct ffmpeg_muxer *stream)
+{
+	if (stream->mux_thread_joinable) {
```
I think you need to call deactivate here as well.

mv2731 replied:
True!

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,339 @@
+{
+	struct encoder_packet *packet;
+	int remaining;
+	bool ret;
```
Initialize ret to true.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,339 @@
+	return NULL;
+}
+
+static int process_packet(struct ffmpeg_muxer *stream)
```
return bool.

mv2731 replied:
Is this referring to the return null line or the fxn signature of process_packet?

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,339 @@
+		if (os_event_try(stream->stop_event) == 0)
+			break;
+
+		int ret = process_packet(stream);
```
ret should be bool.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,339 @@
+	keyint_sec = obs_data_get_int(settings, "keyint_sec");
+	if (keyint_sec)
+		dstr_catf(&stream->muxer_settings, " hls_time=%d", keyint_sec);
+	stream->keyint_sec = keyint_sec;
```
Only set keyint_sec if it's greater than 0.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,339 @@
+
+	buffer_duration_usec = stream->last_dts_usec - first.dts_usec;
+
+	if (buffer_duration_usec > drop_threshold)
```
You need to make sure the units are the same. You are comparing usec with sec.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,339 @@
+
+static void drop_frames(struct ffmpeg_muxer *stream, int highest_priority)
+{
+	int num_frames_dropped = 0;
```
You don't need num_frames_dropped if you are incrementing dropped_frames directly.

mv2731 committed:
obs-ffmpeg: Run clang-format

mv2731 committed:
obs-ffmpeg: Scale drop-frames accordingly

mv2731 committed:
obs-ffmpeg: Format with clang

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,337 @@
+	keyint_sec = obs_data_get_int(settings, "keyint_sec");
+	if (keyint_sec)
+		dstr_catf(&stream->muxer_settings, " hls_time=%d", keyint_sec);
+	if (keyint_sec > 0) {
+		stream->keyint_sec = keyint_sec;
+	}
```
You can combine this with the previous if statement.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,337 @@
+
+	buffer_duration_usec = stream->last_dts_usec - first.dts_usec;
+
+	if (buffer_duration_usec > drop_threshold * pow(10, 6))
```
Just drop_threshold * 1000000

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.h:
```diff
@@ -0,0 +1,52 @@
+	bool is_network;
+	bool threading_buffer;
+
+	int dropped_frames;
+	int min_priority;
+	int64_t last_dts_usec;
```
Group the HLS specific fields together.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,337 @@
+		}
+	}
+
+	stream->active = false;
```
os_atomic_set_bool(&stream->active, false);

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,337 @@
+#include <math.h>
```
Not needed.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,337 @@
+#include <math.h>
+#include <obs-avc.h>
+#include <obs-module.h>
+#include <util/circlebuf.h>
```
Is this needed?

mv2731 committed:
obs-ffmpeg: Clean up pull request

mv2731 committed:
obs-ffmpeg: Reorder the members of ffmpeg_muxer

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,334 @@
+	remaining = stream->mux_packets.num;
+	if (remaining) {
+		packet = &stream->mux_packets.array[0];
+		ret = write_packet(stream, packet);
```
Increment total_bytes by packet->size.

mv2731 replied:
Pretty sure this happens in write_packet? Should I do it in process_packet?

mv2731 committed:
obs-ffmpeg: Alter includes

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -497,6 +395,8 @@ static void signal_failure(struct ffmpeg_muxer *stream)
 
 	ret = deactivate(stream, 0);
+	if (stream->threading_buffer)
+		hls_deactivate(stream, 0);
```
else 
    ret = deactivate(stream, 0);

ushadow replied:
hls_deactivate can return an int like deactivate.

mv2731 committed:
obs-ffmpeg: Let hls_deactive return int

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.h:
```diff
@@ -0,0 +1,60 @@
+	struct dstr printable_path;
+	struct dstr stream_key;
+	int keyint_sec;
+	DARRAY(struct encoder_packet) mux_packets;
+	pthread_t mux_thread;
+	bool mux_thread_joinable;
```
These are used by replay buffer as well. So separate them out. Put these together just like the original file.

DARRAY(struct encoder_packet) mux_packets;
	pthread_t mux_thread;
	bool mux_thread_joinable;
	volatile bool muxing;

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.h:
```diff
@@ -0,0 +1,60 @@
+	int64_t stop_ts;
+	uint64_t total_bytes;
+	struct dstr path;
+	struct dstr muxer_settings;
```
Move muxer_settings to HLS section too.

mv2731 committed:
obs-ffmpeg: Organize HLS fields

mv2731 committed:
obs-ffmpeg: Fix usage of ret in write_thread

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -685,14 +574,15 @@ struct obs_output_info ffmpeg_hls_muxer = {
+	.encoded_packet = ffmpeg_hls_mux_data,
 	.get_total_bytes = ffmpeg_mux_total_bytes,
 	.get_properties = ffmpeg_mux_properties,
 	.get_connect_time_ms = ffmpeg_mpegts_mux_connect_time,
```
Is this needed?

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -685,14 +574,15 @@ struct obs_output_info ffmpeg_hls_muxer = {
-	.encoded_packet = ffmpeg_mux_data,
+	.encoded_packet = ffmpeg_hls_mux_data,
 	.get_total_bytes = ffmpeg_mux_total_bytes,
 	.get_properties = ffmpeg_mux_properties,
```
Is get_properties needed?

mv2731 replied:
I don't think so!

mv2731 committed:
obs-ffmpeg: Run clang-format

mv2731 committed:
obs-ffmpeg: Remove get properties

mv2731 committed:
obs-ffmpeg: Fix spacing in hls-mux.h

mv2731 committed:
obs-ffmpeg: Remove hls_deactivate

mv2731 committed:
obs-ffmpeg: Edit deactivate to include proper print

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -465,10 +355,30 @@ static int deactivate(struct ffmpeg_muxer *stream, int code)
 	os_atomic_set_bool(&stream->stopping, false);
+
+	if (stream->is_hls) {
+
```
Remove extra empty line.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.h:
```diff
@@ -0,0 +1,62 @@
+	volatile bool muxing;
+
+	/* HLS */
+	struct dstr printable_path;
+	struct dstr stream_key;
+	struct dstr muxer_settings;
```
Since printable_path, stream_key and muxer_settings are also used in the obs-ffmpeg-mux.c file, now I think it's better to keep them with first block fields indicating that they are shared.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.h:
```diff
@@ -0,0 +1,62 @@
+	int keyframes;
+	obs_hotkey_id hotkey;
+
+	DARRAY(struct encoder_packet) mux_packets;
```
Add a comment that these are accessed by both replay buffer and HLS.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -118,8 +70,6 @@ static void ffmpeg_mux_destroy(void *data)
 
 	os_process_pipe_destroy(stream->pipe);
 	dstr_free(&stream->path);
-	dstr_free(&stream->printable_path);
-	dstr_free(&stream->stream_key);
```
Keep these two dstr_free.

mv2731 commented:
I responded to all of these comments!

mv2731 committed:
obs-ffmpeg

mv2731 committed:
obs-ffmpeg: Clarify which fields are HLS specific

mv2731 committed:
obs-ffmpeg: Run clang-format

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,319 @@
+	remaining = stream->mux_packets.num;
+	if (remaining) {
+		packet = &stream->mux_packets.array[0];
+		ret = write_packet(stream, packet);
```
I think there is a problem here. We can't hold the lock while calling write_packet because write_packet is blocking while writing to the pipe. The way it is now makes the threading not effective because the encoder thread may not be able to add packet to the dynamic array because it can't get the lock.

ushadow replied:
I tried this, and I got 0 skipped frames!

static bool process_packet(struct ffmpeg_muxer *stream)
{
	struct encoder_packet packet;
	int remaining;
	bool ret = true;

	pthread_mutex_lock(&stream->write_mutex);
	remaining = stream->mux_packets.num;
	if (remaining) {
		packet = stream->mux_packets.array[0];
		da_erase(stream->mux_packets, 0);
	}
	pthread_mutex_unlock(&stream->write_mutex);

	ret = write_packet(stream, &packet);
	obs_encoder_packet_release(&packet);
	return ret;
}

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -453,7 +345,7 @@ static int deactivate(struct ffmpeg_muxer *stream, int code)
 
 		info("Output of file '%s' stopped",
-		     dstr_is_empty(&stream->printable_path)
+		     (dstr_is_empty(&stream->printable_path))
```
This extra parentheses are not necessary.

mv2731 committed:
obs-ffmpeg: Remove blocking call to write_packet

Remove blocking call to write_packet from within
critical section. This introduces an issue where packets
cannot be written to the array because the lock is held
by a blocking write_packet call.

mv2731 committed:
obs-ffmpeg: Fix locking issue in process_pkt

mv2731 committed:
obs-ffmpeg: Change spacing

mv2731 committed:
obs-ffmpeg: Resolve mem_leak from muxer_settings

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,318 @@
+	dstr_copy(&stream->stream_key, stream_key);
+	dstr_copy(&path, path_str);
+	dstr_replace(&path, "{stream_key}", stream_key);
+	dstr_catf(&stream->muxer_settings,
```
Does this result in muxer_settings getting longer and longer over multiple runs?

mv2731 replied:
I tested for that. Interestingly, it doesn't. However, to be sure, we could go back to calling dstr_init to 0 it out but free muxer_settings in deactivate, rather than destroy.

mv2731 committed:
obs-ffmpeg: Resolve dstr_catf muxer_settings

mv2731 committed:
obs-ffmpeg: Replace mux_packets with circlebuf

Eliminate unnecessary complexity from using
a dynamic array for storing packets rather than
a circlebuf.

mv2731 committed:
obs-ffmpeg: Run clang-format

mv2731 committed:
obs-ffmpeg: Edit destroy

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,332 @@
+void *ffmpeg_hls_mux_create(obs_data_t *settings, obs_output_t *output)
+{
+	struct ffmpeg_muxer *stream = bzalloc(sizeof(*stream));
+	pthread_mutex_init_value(&stream->write_mutex);
```
I'm wondering if this is needed.

mv2731 replied:
the pthread_mutex_init_value call? Yeah I agree it seems strange/redundant but I observed it in almost every file where mutexes are used and thought there must be a reason. It's defined in threading.h which might shine some light.

mv2731 replied:
![image](https://user-images.githubusercontent.com/60421398/90040436-e648f400-dc95-11ea-9d5f-bd2bf9d8e5dc.png)
It dereferences our mutex ptr and stores a mutex set to PTHREAD_MUTEX_INITIALIZER in the memory location pointed to by our mutex ptr.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,332 @@
+		if (!ret) {
+			int code = OBS_OUTPUT_ERROR;
+			obs_output_signal_stop(stream->output, code);
+			deactivate(stream, 0);
```
obs_output_signal_stop(stream->output, OBS_OUTPUT_ERROR);

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,332 @@
+
+	if (stream->min_priority < highest_priority)
+		stream->min_priority = highest_priority;
+	if (!num_frames_dropped)
+		return;
```
I think these two lines are not necessary.

mv2731 replied:
I agree. I think all that gets achieved by leaving it in is returning one step faster, but I don't think the adding operation will be super expensive.

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,332 @@
+	int64_t buffer_duration_usec;
+	int priority = pframes ? OBS_NAL_PRIORITY_HIGHEST
+			       : OBS_NAL_PRIORITY_HIGH;
+	int64_t drop_threshold = 2 * stream->keyint_sec;
```
I think we need to handle the case when keyint_sec may be zero. So we need a default threshold.

ushadow replied:
drop_threshold_sec

mv2731 committed:
obs-ffmpeg: Make initial changes based on review

mv2731 committed:
obs-ffmpeg: Add debug comments

mv2731 committed:
obs-ffmpeg: Remove unnecessary prints

mv2731 committed:
obs-ffmpeg: Remove print statements and clang-format

mv2731 committed:
obs-ffmpeg: Order obs-ffmpeg sources

mv2731 committed:
obs-ffmpeg: Reorder ffmpeg_muxer elements

mv2731 committed:
obs-ffmpeg: Clean up obs-ffmpeg-hls-mux.c

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,336 @@
+int hls_stream_dropped_frames(void *data)
+{
+	struct ffmpeg_muxer *stream = data;
+	return stream->dropped_frames;
```
Will be good to protect dropped_frames with mutex since it can be accessed by different threads. pthread_mutex_lock(&stream->write_mutex);

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,336 @@
+			break;
+
+		bool ret = process_packet(stream);
+		if (!ret) {
```
if (!process_packet(stream))

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,336 @@
+	/* write headers and start capture */
+	os_atomic_set_bool(&stream->active, true);
+	os_atomic_set_bool(&stream->capturing, true);
+	os_atomic_set_bool(&stream->is_hls, true);
```
stream->is_hls = true

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -465,10 +358,31 @@ static int deactivate(struct ffmpeg_muxer *stream, int code)
 }
 
-static void ffmpeg_mux_stop(void *data, uint64_t ts)
+void ffmpeg_mux_stop(void *data, uint64_t ts)
```
What if ts != 0? We may need a custom stop function for hls.

mv2731 replied:
I'll look into this

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,336 @@
+	}
+
+	if (stopping(stream)) {
+		if (packet->sys_dts_usec >= stream->stop_ts) {
```
I'm wondering if sys_dts_usec is actually set. Could you log this value and check?

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -465,10 +358,31 @@ static int deactivate(struct ffmpeg_muxer *stream, int code)
 
 	os_atomic_set_bool(&stream->stopping, false);
+
+	if (stream->is_hls) {
```
I think we better write a different deactivate function for hls. I think we can't just destroy the pipe like this because the write thread is still running and it may write to the pipe.

mv2731 replied:
switched the order for stream->is_hls

mv2731 committed:
obs-ffmpeg: Print system_dts_usec for comparison

mv2731 committed:
obs-ffmpeg: Reorder deactivate

mv2731 committed:
obs-ffmpeg: Run clang-format

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,337 @@
+		}
+	}
+
+	os_atomic_set_bool(&stream->active, false);
```
We can move

obs_output_signal_stop(stream->output,
					       OBS_OUTPUT_ERROR);
deactivate(stream, 0);

above here and delete os_atomic_set_bool(&stream->active, false);

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-mux.c:
```diff
@@ -607,6 +521,8 @@ static void ffmpeg_mux_data(void *data, struct encoder_packet *packet)
 
 	if (stopping(stream)) {
 		if (packet->sys_dts_usec >= stream->stop_ts) {
+			printf("\n* * * ffmpeg_mux_data: packet->sys_dts_usec and stream->stop_ts: %d, %d * * *\n",
```
Need to print in ffmpeg_hls_mux_data.

mv2731 replied:
oh yep! of course. Didn't finish testing here

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,337 @@
+
+	while (os_sem_wait(stream->write_sem) == 0) {
+		if (os_event_try(stream->stop_event) == 0)
+			break;
```
return instead of break

mv2731 replied:
return NULL; ok!

mv2731 committed:
obs-ffmpeg: Move print of sys_dts_usec to hls

mv2731 committed:
obs-ffmpeg: Return after stop_event found

mv2731 committed:
obs-ffmpeg: Remove mutex in hls_stream_dropped_frames

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,338 @@
+			return NULL;
+
+		if (!process_packet(stream)) {
+			obs_output_signal_stop(stream->output,
+					       OBS_OUTPUT_ERROR);
+			deactivate(stream, 0);
```
We don't need these any more once they are moved out of the while loop.

mv2731 committed:
obs-ffmpeg: Remove print statement of sys_dts_usec

ushadow commented on plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c:
```diff
@@ -0,0 +1,333 @@
+		if (os_event_try(stream->stop_event) == 0)
+			return NULL;
+
+		if (!process_packet(stream)) {
```
The curly braces are unnecessary.

mv2731 replied:
Good point!

ushadow reviewed (changes requested):
See comment.

mv2731 committed:
obs-ffmpeg: Remove curly braces

mv2731 committed:
obs-ffmpeg: Run clang-format
